lib/Compiler.js

```diff
@@ -26,38 +26,91 @@ const RequestShortener = require("./RequestShortener");
 const { makePathsRelative } = require("./util/identifier");
 const ConcurrentCompilationError = require("./ConcurrentCompilationError");
 
+/**
+ * @typedef {Object} CompilationParams
+ * @property {NormalModuleFactory} normalModuleFactory
+ * @property {ContextModuleFactory} contextModuleFactory
+ * @property {Set<any>} compilationDependencies
+ */
+
+/** @typedef {string | string[] | {[entryName: string]: string} | EntryTypeFn } EntryTypesVals */
+/** @typedef {() => EntryTypesVals} EntryTypeFn */
+
 class Compiler extends Tapable {
 	constructor(context) {
 		super();
 		this.hooks = {
+			/** @type {SyncBailHook<Compilation>} */
 			shouldEmit: new SyncBailHook(["compilation"]),
+
+			/** @type {AsyncSeriesHook<Stats>} */
 			done: new AsyncSeriesHook(["stats"]),
+
+			/** @type {AsyncSeriesHook<>} */
```

Comment:
This is odd. We should decide how to declare _empty_ hooks.

1. `AsyncSeriesHook`
1. `AsyncSeriesHook<>`
1. `AsyncSeriesHook<void>`

Reply:
Yeah I agree, I think I just got inconsistent after 100+ types, I'm fine with `AsyncSeriesHook` personally.

Reply:
We can't used `Hook<void>` especially for AsyncHook's because this will remove the callback parameter as being valid.